### PR TITLE
Add some more tests for DEFEXPI functionality

### PR DESCRIPTION
**Pull request by ecpeterson**
cf. https://github.com/rigetti/quilc/pull/498#pullrequestreview-334160910

### CODE REVIEW

File: src/parser.lisp [hunk 1]
```diff
@@ -984,20 +984,22 @@ If ENSURE-VALID is T, then a memory reference such as 'foo[0]' will result in an
                        :arguments qubit-list))))
 
 (defun pauli-term->matrix (term arguments parameters parameter-names)
-  (let* ((prefactor-fn
```
Comment (ecpeterson) [inline src/parser.lisp]: I was wondering where this commit snuck off to.

**Comment by ecpeterson** [comment]
The only suggested test not added is an explicit test involving `parametric-diagonal-compiler`. The role of that compiler is to decompose exponentiated Pauli sums only containing `Z`s into `CNOT`s and `RZ`s (in a way that permits the Pauli terms to have underdetermined coefficients). I'm honestly not sure what a good test would look like that would instill any more confidence in me (or provide a more isolated bug report) than `test-parametric-defexpi`.

**Comment by jmbr** [comment]
Suggestion: test against `cl-quil.clifford:exp-pauli`.

File: tests/translator-tests.lisp [hunk 2]
```diff
@@ -373,6 +373,54 @@ U(time)~{ ~a~}"
...
+PRAGMA INITIAL_REWIRING \"NAIVE\"
+DEFGATE EXPI(%beta) p AS PAULI-SUM:
+    Z(%beta/2) p
+
+EXPI(2.0) 0
+RZ(-2.0) 0"))
+      (let* ((cpp (compiler-hook (parse-quil prog-text) chip))
+             (m (parsed-program-to-logical-matrix cpp)))
+        (is (quil::double= 1d0 (abs (magicl:ref m 0 0))))
+        (loop :for j :below (magicl:matrix-rows m)
```
Comment (notmgsk) [inline tests/translator-tests.lisp]: `magicl:identityp`?

Comment (ecpeterson) [reply tests/translator-tests.lisp]: This is checking `identityp` up to phase.

File: src/parser.lisp [hunk 3]
```diff
@@ -984,24 +984,28 @@ If ENSURE-VALID is T, then a memory reference such as 'foo[0]' will result in an
...
            (typecase (pauli-term-prefactor term)
              (number
-              (lambda (&rest args) (declare (ignore args)) (pauli-term-prefactor term)))
+              (pauli-term-prefactor term))
              (delayed-expression
-              (compile nil `(lambda ,parameter-names
-                              (declare (ignorable ,@parameter-names))
-                              ,(delayed-expression-expression (pauli-term-prefactor term)))))
+              (apply
+               (compile nil `(lambda ,parameter-names
```
Comment (stylewarning) [inline src/parser.lisp]: Definitely not for this commit, but we might instead simply interpret the expression instead of calling the compiler.

**Comment by stylewarning** [comment]
> Suggestion: test against `cl-quil.clifford:exp-pauli`.

i think this is a good suggestion. Could @jmbr or @ecpeterson make a "good-first-issue" for this?

**Comment by ecpeterson** [comment]
I will.

**Pull request merged**